apps/backend/src/auth/google.auth.controller.ts

```diff
@@ -0,0 +1,112 @@
+import Elysia, { t } from "elysia";
+import { googleAuth } from "../lucia";
+import cookie from "@elysiajs/cookie";
+import { jwt } from "../plugins";
+import { UnauthorizedError } from "../errors";
+import { OAuthRequestError } from "@lucia-auth/oauth";
+import { randomUUID as uuidv4 } from "crypto";
+
+const schemaDetail = {
+  tags: ["Auth"],
+};
+
+export default new Elysia({ prefix: "/google" })
+  .use(cookie())
+  .use(jwt)
+  .error({
+    UnauthorizedError,
+  })
+  .get(
+    "/authorize",
+    async ({ set, setCookie }) => {
+      const [authUrl, authState] = await googleAuth.getAuthorizationUrl();
+      setCookie("googleAuthState", authState, { maxAge: 60 });
+      set.redirect = authUrl.toString();
+    },
+    {
+      detail: {
+        ...schemaDetail,
+        responses: {
+          302: {
+            description: "Redirect to Google authorization page",
+          },
+        },
+      },
+    },
+  )
+  .get(
+    "/callback",
+    async ({
+      set,
+      query: { code, state },
+      cookie: { googleAuthState },
+      jwt,
+    }) => {
+      if (!googleAuthState) {
+        throw new UnauthorizedError(`Missing cookie 'googleAuthState'`);
+      }
+      if (googleAuthState.toString() !== state) {
+        throw new UnauthorizedError(
+          "The 'state' query param doesn't match the 'googleAuthState' cookie",
+        );
+      }
+
+      try {
+        const { getExistingUser, googleUser, createUser } =
+          await googleAuth.validateCallback(code);
+
+        const getUser = async () => {
+          const existingUser = await getExistingUser();
+          if (existingUser) return existingUser;
+
+          return createUser({
+            userId: uuidv4(),
+            attributes: {
+              username: googleUser.name,
+              email: googleUser.email,
+            },
+          });
+        };
```

Comment:
lucia doesn't provide any get or create methods?

Reply:
Nope, we'd have to create custom (preferrably generic) ones. Lucia only provides `getExistingUser`, `createUser`, `createKey` and the provider-specific User API response.

Reply:
`google.auth.controller.ts` and `github.auth.controller.ts` have to get refactored into a more generic external oauth flow, because they are almost identical in code structure.